numpyro/infer/util.py

```diff
@@ -673,17 +674,47 @@ def _predictive(
     posterior_samples,
     batch_shape,
     return_sites=None,
+    infer_discrete=False,
     parallel=True,
     model_args=(),
     model_kwargs={},
 ):
-    model = numpyro.handlers.mask(model, mask=False)
+    masked_model = numpyro.handlers.mask(model, mask=False)
+    if infer_discrete:
+        # inspect the model to get some structure
+        rng_key, subkey = random.split(rng_key)
+        batch_ndim = len(batch_shape)
+        prototype_sample = tree_map(
+            lambda x: jnp.reshape(x, (-1,) + jnp.shape(x)[batch_ndim:])[0],
+            posterior_samples,
+        )
+        prototype_trace = trace(
+            seed(substitute(masked_model, prototype_sample), subkey)
+        ).get_trace(*model_args, **model_kwargs)
+        first_available_dim = -_guess_max_plate_nesting(prototype_trace) - 1
 
     def single_prediction(val):
         rng_key, samples = val
-        model_trace = trace(seed(substitute(model, samples), rng_key)).get_trace(
-            *model_args, **model_kwargs
-        )
+        if infer_discrete:
+            from numpyro.contrib.funsor import config_enumerate
+            from numpyro.contrib.funsor.discrete import _sample_posterior
+
+            model_trace = prototype_trace
+            temperature = 1
+            pred_samples = _sample_posterior(
+                config_enumerate(condition(model, samples)),
```

Comment:
I'm surprised you're automatically configuring the model for enumeration here.  In Pyro we let the user decide which variables are enumerated.  That way a guide might sample some discrete latents, and the model might enumerate others.  I would expect `Predictive` to use the guide's samples if provided, and only enumerate sites already marked for enumeration by the user.

I guess if in NumPyro you automatically wrap models with `@config_enumerate` inside MCMC, then it would also make sense to automatically wrap them here.  Still, it would be nice to support SVI with guides that sample some or all discrete latent variables.

Reply:
> let the user decide which variables are enumerated

We'll aim for this except for algorithms that do not work with discrete latent sites (in this case, it makes sense to enable enumeration by default and raise errors for invalid models). Currently, DiscreteHMCGibbs will perform Gibbs update for discrete latent sites that are not marked "enumerated". Then posterior samples will include all latent variables except for those marked "enumerated".

When `infer_discrete` is used, I assumed that the latent sites that are not available in the `posterior_samples` or `guide` are enumerated (those latent sites are the `samples` variable in the above code). So it makes sense to me to `config_enumerate` them by default (`config_enumerate` will skip observed sites, including those sites in `samples`). What do you think? (it seems to not contradict with the usage case in your comment)